telethon/client/telegrambaseclient.py

```diff
@@ -654,6 +654,21 @@ def set_proxy(self: 'TelegramClient', proxy: typing.Union[tuple, dict]):
             else:
                 connection._proxy = proxy
 
+    def _save_states_and_entities(self: 'TelegramClient'):
+        entities = self._mb_entity_cache.get_all_entities()
+
+        # Piggy-back on an arbitrary TL type with users and chats so the session can understand to read the entities.
+        # It doesn't matter if we put users in the list of chats.
+        self.session.process_entities(types.contacts.ResolvedPeer(None, [e._as_input_peer() for e in entities], []))
+
+        ss, cs = self._message_box.session_state()
+        self.session.set_update_state(0, types.updates.State(**ss, unread_count=0))
+        now = datetime.datetime.now()  # any datetime works; channels don't need it
+        entities_ids = [e.id for e in entities]
+        for channel_id, pts in cs.items():
+            if channel_id in entities_ids:
```

Comment:
I would like an `else` statement which logs either an information message or perhaps even a warning when a channel state was *not* saved. Silently dropping the update state for a channel doesn't feel very nice.

Reply:
But the channel is silently not added to entities if it was received as min. I can add else, but when the channel is not added to the entities due to the min state, we will receive flood to the logs every minute.

Reply:
In my opinion, silently not adding a channel to entities when there's no `access_hash` that can be used forever after is fine. (And `min` indeed can't be used whenever and however we want.) The entity cache is a "nice to have" after all, so you can use IDs to refer to entities without needing to store the hash yourself. And it's provided as a best-effort, not a guarantee.

But silently dropping the update state means `catch_up` won't work as expected as it will be missing some channels. So I think we should log when an update state is not saved, but there is no need to do so when a channel won't be present in cache.

With that said, I find it weird that we get updates for channels but don't actually have the real `access_hash` for them. Do you know when this happens? Is there a way for me to reproduce that behavior?

I think we should fix this problem at the root, rather than try to work around it.

Reply:
I caught this problem several times and caught that the update comes, but the channel is in the min state and it is dropped, but the state is recorded and at the next start the catch_up tries to pull the channel from entities, but it is not there.
Tomorrow I will try to catch this issue again and provide you with an update that causes the situation.
I can cut this piece out of this pr to deal with this problem separately

Reply:
Yes, please do that.

Note that if you delete the `if` we can also get rid of `entities_ids = {e.id for e in entities}` as it will no longer be used.